### ISSUE TITLE
Error "Cannot read property 'param' of undefined" when performing transaction with latest ethereumjs/tx lib

### ISSUE THREAD

**Issue by rmeissner** [nomiclabs/hardhat#1537]
Edit: see [this comment](https://github.com/nomiclabs/hardhat/issues/1537#issuecomment-859688275) for a possible fix.

---

When using the latest version of all libs I get an error when executing a tx:

```
     TypeError: Cannot read property 'param' of undefined
      at FakeSenderTransaction.getDataFee (node_modules/@ethereumjs/tx/src/baseTransaction.ts:145:36)
      at FakeSenderTransaction.getBaseFee (node_modules/@ethereumjs/tx/src/baseTransaction.ts:134:22)
      at VM._runTx (node_modules/@ethereumjs/vm/lib/runTx.ts:222:22)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at runNextTicks (node:internal/process/task_queues:65:3)
      at listOnTimeout (node:internal/timers:526:9)
      at processTimers (node:internal/timers:500:7)
      at VM.runTx (node_modules/@ethereumjs/vm/lib/runTx.ts:168:20)
      at HardhatNode._runTxAndRevertMutations (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:1687:14)
      at HardhatNode._runInPendingBlockContext (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:1520:14)
``` 

After adding
```json
  "resolutions": {
    "@ethereumjs/vm": "5.3.2",
    "@ethereumjs/tx": "3.1.3"
   }
``` 
It worked again.

Not sure if this should go to the `ethereumjs` repo, let me know if I should move it.

For reference this is the latest release on ethereumtx (was 3h prior to posting this):
https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Ftx%403.2.1

**Comment by fvictorio** [comment]
Ugh, that sucks, thanks for letting us know. We are going to make a release next week, so we might use a fixed version for the ethereumjs dependencies to prevent this from happening again.

**Comment by holgerd77** [comment]
Hi Richard, could you post some code what you have to run on the Hardhat side to reproduce this?

**Comment by holgerd77** [comment]
Yeah, I think I am actually seeing what is happening. This is from the [FakeSenderTransaction](https://github.com/nomiclabs/hardhat/blob/116c8994e272592b6bfbec3ad1424738ae0d0113/packages/hardhat-core/src/internal/hardhat-network/provider/transactions/FakeSenderTransaction.ts#L91) constructor implementation.

The `Common` instantiation on the latest release moved from being done in the `BaseTransaction` constructor to the subclass constructors, e.g. for `EIP1559FeeMarketTransaction` see [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/eip1559Transaction.ts#L185).

Since you are directly inheriting from `BaseTransaction` this instantiation is now omitted and `Common` is not present. This is the "Cannot read property 'param' of undefined" error.

I generally have to say that this "inheriting from BaseTransaction" logic for "FakeSenderTransaction" is likely not meant to be stable. You are omitting all sorts of logic from the respective implementation classes with this and chances that this will break again are high. Maybe we can work together here and reintroduce some sort of build in FakeTransaction class or something which is more robust to use.

**Comment by holgerd77** [comment]
Or - might also be useful - the respective PR on our side directly: https://github.com/ethereumjs/ethereumjs-monorepo/pull/1292

**Comment by fvictorio** [comment]
If anyone runs into this issue and is using yarn, add this to your `package.json`:

```
  "resolutions": {
    "@ethereumjs/block": "3.2.1",
    "@ethereumjs/blockchain": "5.2.1",
    "@ethereumjs/common": "2.2.0",
    "@ethereumjs/tx": "3.1.4",
    "@ethereumjs/vm": "5.3.1"
   }
```

and then run `yarn` again. If that doesn't fixes it, please let me know.

**Comment by starlabs007** [comment]
Note to others if you're using yarn workspaces, the resolutions block needs to be placed in the **root** `package.json`.

**Comment by fvictorio** [comment]
I just published an RC version that should fix this. If you can, please try it out by doing `yarn add --dev hardhat@rc` or `npm i --save-dev hardhat@rc` and let me know if it fixes the problem (or not).

**Comment by fvictorio** [comment]
We published [Hardhat v2.3.3](https://github.com/nomiclabs/hardhat/releases/tag/hardhat-core-v2.3.3), which should fix this problem. Please re-open this issue or let us know if you keep running into this problem after upgrading to it.
